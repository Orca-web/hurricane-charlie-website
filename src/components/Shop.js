import React, { Component } from 'react';
import ShopItem from './ShopItem';
import shopItems from '../globals/shopItems';

function validateEmail(email) {
  const emailRegx = /^(([^<>()\[\]\\.,;:\s@"]+(\.[^<>()\[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/;
  if (email.length === 0) {
    return "email is required";
  } else if (!email.match(emailRegx)) {
    return "incorrect email format";
  }
  return null
}

class Shop extends Component {
  constructor() {
    super();
    this.state = { email: '', msg: null };
    this.mailListChange = this.mailListChange.bind(this);
    this.mailListSubmit = this.mailListSubmit.bind(this);
  }

  async mailListSubmit() {
    const email = { email: this.state.email };
    let msg = validateEmail(email.email);

    if (!msg) {
      await fetch(process.env.MAILER_URL, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json'
        },
        body: JSON.stringify(email)
      })
      .then(resp => {
        if (resp.ok) msg = 'You\'ve been added to the mailing list';
      })
      .catch(err => {
        msg = 'Something went wrong, please try again.'
      });
    }
    this.setState({msg})
  }

  componentDidMount() {
    window.scrollTo(0, 0)
  }

  mailListChange(e) {
    const email = e.target.value; 
    this.setState({ email })
  }

  render() {
    const statusMsg = this.state.msg ? <p>{ this.state.msg }</p> : null;
    return (
      <div>
        
        <div className="shop-container">
          {shopItems.map((itemData) => (
            <ShopItem key={ itemData.id } product={ itemData }/>
          ))}
        </div>

        <div className="mailer-cta">
          <h2>Be the first to know about new items</h2>
          <div className="mailer-form">
            { statusMsg }
            <input 
              value={ this.state.email } 
              onChange={ this.mailListChange } 
              type="email" placeholder="your email" 
            />
            <button onClick={ this.mailListSubmit }>join</button>
          </div>
        </div>

      </div>
    );
  }
}

export default Shop;