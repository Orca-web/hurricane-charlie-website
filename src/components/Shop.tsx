import React, { Component } from 'react';
import ShopItem from './ShopItem';
import { shopItems } from '../globals/shopItems';
import { emailCheck } from '../helpers/validators';

import imageImport from '../helpers/imageImport';

const images = imageImport(require.context('../images/shop', false, /\.(png|jpe?g|svg)$/));

function validateEmail(email: string): string | null {
  if (email.length === 0) {
    return "email is required";
  } else if (!emailCheck(email)) {
    return "incorrect email format";
  }
  return null
}

class Shop extends Component {
    state = {
      email: '',
      msg: null,
    };

  componentDidMount() {
    window.scrollTo(0, 0)
  }

  mailListSubmit = async (): Promise<void> => {
    const email = { email: this.state.email };
    let msg = validateEmail(email.email);

    if (!msg) {
      try {
        await fetch(process.env.MAILER_URL!, {
          method: 'POST',
          headers: {
            'Content-Type': 'application/json'
          },
          body: JSON.stringify(email)
        })
        msg = 'You\'ve been added to the mailing list';
      } catch (error) {
        msg = 'Something went wrong, please try again.';
      }
    }
    this.setState({msg})
  }

  mailListChange(e: React.FormEvent<HTMLInputElement>): void {
    const email = e.currentTarget.value; 
    this.setState({ email })
  }

  render() {
    const statusMsg = this.state.msg ? <p>{ this.state.msg }</p> : null;
    return (
      <div>
        <div className="shop-container">
          {shopItems.map((itemData) => (
            <ShopItem key={ itemData.id } product={ itemData } imgFile={images[itemData.url]}/>
          ))}
        </div>

        <div className="mailer-cta">
          <h2>Be the first to know about new items</h2>
          { statusMsg }
          <div className="mailer-form">
            <input 
              value={ this.state.email } 
              onChange={ this.mailListChange } 
              type="email" placeholder="your email" 
            />
            <button onClick={ this.mailListSubmit }>join</button>
          </div>
        </div>
      </div>
    );
  }
}

export default Shop;