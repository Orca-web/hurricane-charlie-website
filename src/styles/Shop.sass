.mailer-cta
  background: lime
  display: flex
  flex-direction: column
  padding: 50px 20px
  text-align: center

  h2
    margin: 0 20px 10px

.shop-item
  background: pink
  padding: 10px

  img
    width: 200px

.mailer-form
  input, button
    background: none
    padding: 10px 20px
    border: 3px solid black
    font-size: inherit
    outline: none
    
  input
    margin-right: 20px
    &::placeholder
      font-family: serif
      font-style: italic
      color: black

  button
    cursor: pointer
    transition: color 0.3s ease-in, background 0.3s ease-in
    &:hover
      background: black
      color: lime