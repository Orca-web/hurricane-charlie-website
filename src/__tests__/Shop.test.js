import React from 'react';
import { shallow } from 'enzyme';
import Shop from '../components/Shop';

const app = shallow(<Shop />);

describe('Shop', () => {
  it('renders correctly', () => {
    expect(app).toMatchSnapshot();
  });
});