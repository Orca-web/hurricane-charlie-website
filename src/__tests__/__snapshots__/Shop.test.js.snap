// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Shop renders correctly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Shop />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <div
          className="shop-container"
        >
          <ShopItem
            product={
              Object {
                "desc": "monochrome screenprint",
                "id": 1,
                "price": 9.95,
                "shopUrl": "mushroomjack",
                "soldOut": false,
                "title": "Mushroom Lumberjack",
                "url": "shop01.jpg",
              }
            }
          />
          <ShopItem
            product={
              Object {
                "desc": "five colour screenprint",
                "id": 2,
                "price": 22.5,
                "shopUrl": "patterned-beetle",
                "soldOut": false,
                "title": "Patterned Beetle",
                "url": "shop02.jpg",
              }
            }
          />
          <ShopItem
            product={
              Object {
                "desc": "glossy giclee full colour",
                "id": 3,
                "price": 3.95,
                "shopUrl": "desert-smoker",
                "soldOut": false,
                "title": "Desert Smoker",
                "url": "shop03.jpg",
              }
            }
          />
          <ShopItem
            product={
              Object {
                "desc": "art stock giclee",
                "id": 4,
                "price": 12.95,
                "shopUrl": "robot-workshop",
                "soldOut": false,
                "title": "Robot Workshop",
                "url": "shop04.jpg",
              }
            }
          />
          <ShopItem
            product={
              Object {
                "desc": "five colour screenprint",
                "id": 5,
                "price": 22.5,
                "shopUrl": "patterned-butterfly",
                "soldOut": false,
                "title": "Patterned Butterfly",
                "url": "shop05.jpg",
              }
            }
          />
          <ShopItem
            product={
              Object {
                "desc": "glossy giclee full colour",
                "id": 6,
                "price": 3.95,
                "shopUrl": "desert-portrait",
                "soldOut": true,
                "title": "Desert Portrait",
                "url": "shop06.jpg",
              }
            }
          />
        </div>,
        <div
          className="mailer-cta"
        >
          <h2>
            Be the first to know about new items
          </h2>
          <div
            className="mailer-form"
          >
            <input
              onChange={[Function]}
              placeholder="your email"
              type="email"
              value=""
            />
            <button
              onClick={[Function]}
            >
              join
            </button>
          </div>
        </div>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <ShopItem
              product={
                Object {
                  "desc": "monochrome screenprint",
                  "id": 1,
                  "price": 9.95,
                  "shopUrl": "mushroomjack",
                  "soldOut": false,
                  "title": "Mushroom Lumberjack",
                  "url": "shop01.jpg",
                }
              }
            />,
            <ShopItem
              product={
                Object {
                  "desc": "five colour screenprint",
                  "id": 2,
                  "price": 22.5,
                  "shopUrl": "patterned-beetle",
                  "soldOut": false,
                  "title": "Patterned Beetle",
                  "url": "shop02.jpg",
                }
              }
            />,
            <ShopItem
              product={
                Object {
                  "desc": "glossy giclee full colour",
                  "id": 3,
                  "price": 3.95,
                  "shopUrl": "desert-smoker",
                  "soldOut": false,
                  "title": "Desert Smoker",
                  "url": "shop03.jpg",
                }
              }
            />,
            <ShopItem
              product={
                Object {
                  "desc": "art stock giclee",
                  "id": 4,
                  "price": 12.95,
                  "shopUrl": "robot-workshop",
                  "soldOut": false,
                  "title": "Robot Workshop",
                  "url": "shop04.jpg",
                }
              }
            />,
            <ShopItem
              product={
                Object {
                  "desc": "five colour screenprint",
                  "id": 5,
                  "price": 22.5,
                  "shopUrl": "patterned-butterfly",
                  "soldOut": false,
                  "title": "Patterned Butterfly",
                  "url": "shop05.jpg",
                }
              }
            />,
            <ShopItem
              product={
                Object {
                  "desc": "glossy giclee full colour",
                  "id": 6,
                  "price": 3.95,
                  "shopUrl": "desert-portrait",
                  "soldOut": true,
                  "title": "Desert Portrait",
                  "url": "shop06.jpg",
                }
              }
            />,
          ],
          "className": "shop-container",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": "1",
            "nodeType": "function",
            "props": Object {
              "product": Object {
                "desc": "monochrome screenprint",
                "id": 1,
                "price": 9.95,
                "shopUrl": "mushroomjack",
                "soldOut": false,
                "title": "Mushroom Lumberjack",
                "url": "shop01.jpg",
              },
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": "2",
            "nodeType": "function",
            "props": Object {
              "product": Object {
                "desc": "five colour screenprint",
                "id": 2,
                "price": 22.5,
                "shopUrl": "patterned-beetle",
                "soldOut": false,
                "title": "Patterned Beetle",
                "url": "shop02.jpg",
              },
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": "3",
            "nodeType": "function",
            "props": Object {
              "product": Object {
                "desc": "glossy giclee full colour",
                "id": 3,
                "price": 3.95,
                "shopUrl": "desert-smoker",
                "soldOut": false,
                "title": "Desert Smoker",
                "url": "shop03.jpg",
              },
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": "4",
            "nodeType": "function",
            "props": Object {
              "product": Object {
                "desc": "art stock giclee",
                "id": 4,
                "price": 12.95,
                "shopUrl": "robot-workshop",
                "soldOut": false,
                "title": "Robot Workshop",
                "url": "shop04.jpg",
              },
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": "5",
            "nodeType": "function",
            "props": Object {
              "product": Object {
                "desc": "five colour screenprint",
                "id": 5,
                "price": 22.5,
                "shopUrl": "patterned-butterfly",
                "soldOut": false,
                "title": "Patterned Butterfly",
                "url": "shop05.jpg",
              },
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": "6",
            "nodeType": "function",
            "props": Object {
              "product": Object {
                "desc": "glossy giclee full colour",
                "id": 6,
                "price": 3.95,
                "shopUrl": "desert-portrait",
                "soldOut": true,
                "title": "Desert Portrait",
                "url": "shop06.jpg",
              },
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
        ],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <h2>
              Be the first to know about new items
            </h2>,
            <div
              className="mailer-form"
            >
              <input
                onChange={[Function]}
                placeholder="your email"
                type="email"
                value=""
              />
              <button
                onClick={[Function]}
              >
                join
              </button>
            </div>,
          ],
          "className": "mailer-cta",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Be the first to know about new items",
            },
            "ref": null,
            "rendered": "Be the first to know about new items",
            "type": "h2",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                null,
                <input
                  onChange={[Function]}
                  placeholder="your email"
                  type="email"
                  value=""
                />,
                <button
                  onClick={[Function]}
                >
                  join
                </button>,
              ],
              "className": "mailer-form",
            },
            "ref": null,
            "rendered": Array [
              null,
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "onChange": [Function],
                  "placeholder": "your email",
                  "type": "email",
                  "value": "",
                },
                "ref": null,
                "rendered": null,
                "type": "input",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "join",
                  "onClick": [Function],
                },
                "ref": null,
                "rendered": "join",
                "type": "button",
              },
            ],
            "type": "div",
          },
        ],
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            className="shop-container"
          >
            <ShopItem
              product={
                Object {
                  "desc": "monochrome screenprint",
                  "id": 1,
                  "price": 9.95,
                  "shopUrl": "mushroomjack",
                  "soldOut": false,
                  "title": "Mushroom Lumberjack",
                  "url": "shop01.jpg",
                }
              }
            />
            <ShopItem
              product={
                Object {
                  "desc": "five colour screenprint",
                  "id": 2,
                  "price": 22.5,
                  "shopUrl": "patterned-beetle",
                  "soldOut": false,
                  "title": "Patterned Beetle",
                  "url": "shop02.jpg",
                }
              }
            />
            <ShopItem
              product={
                Object {
                  "desc": "glossy giclee full colour",
                  "id": 3,
                  "price": 3.95,
                  "shopUrl": "desert-smoker",
                  "soldOut": false,
                  "title": "Desert Smoker",
                  "url": "shop03.jpg",
                }
              }
            />
            <ShopItem
              product={
                Object {
                  "desc": "art stock giclee",
                  "id": 4,
                  "price": 12.95,
                  "shopUrl": "robot-workshop",
                  "soldOut": false,
                  "title": "Robot Workshop",
                  "url": "shop04.jpg",
                }
              }
            />
            <ShopItem
              product={
                Object {
                  "desc": "five colour screenprint",
                  "id": 5,
                  "price": 22.5,
                  "shopUrl": "patterned-butterfly",
                  "soldOut": false,
                  "title": "Patterned Butterfly",
                  "url": "shop05.jpg",
                }
              }
            />
            <ShopItem
              product={
                Object {
                  "desc": "glossy giclee full colour",
                  "id": 6,
                  "price": 3.95,
                  "shopUrl": "desert-portrait",
                  "soldOut": true,
                  "title": "Desert Portrait",
                  "url": "shop06.jpg",
                }
              }
            />
          </div>,
          <div
            className="mailer-cta"
          >
            <h2>
              Be the first to know about new items
            </h2>
            <div
              className="mailer-form"
            >
              <input
                onChange={[Function]}
                placeholder="your email"
                type="email"
                value=""
              />
              <button
                onClick={[Function]}
              >
                join
              </button>
            </div>
          </div>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <ShopItem
                product={
                  Object {
                    "desc": "monochrome screenprint",
                    "id": 1,
                    "price": 9.95,
                    "shopUrl": "mushroomjack",
                    "soldOut": false,
                    "title": "Mushroom Lumberjack",
                    "url": "shop01.jpg",
                  }
                }
              />,
              <ShopItem
                product={
                  Object {
                    "desc": "five colour screenprint",
                    "id": 2,
                    "price": 22.5,
                    "shopUrl": "patterned-beetle",
                    "soldOut": false,
                    "title": "Patterned Beetle",
                    "url": "shop02.jpg",
                  }
                }
              />,
              <ShopItem
                product={
                  Object {
                    "desc": "glossy giclee full colour",
                    "id": 3,
                    "price": 3.95,
                    "shopUrl": "desert-smoker",
                    "soldOut": false,
                    "title": "Desert Smoker",
                    "url": "shop03.jpg",
                  }
                }
              />,
              <ShopItem
                product={
                  Object {
                    "desc": "art stock giclee",
                    "id": 4,
                    "price": 12.95,
                    "shopUrl": "robot-workshop",
                    "soldOut": false,
                    "title": "Robot Workshop",
                    "url": "shop04.jpg",
                  }
                }
              />,
              <ShopItem
                product={
                  Object {
                    "desc": "five colour screenprint",
                    "id": 5,
                    "price": 22.5,
                    "shopUrl": "patterned-butterfly",
                    "soldOut": false,
                    "title": "Patterned Butterfly",
                    "url": "shop05.jpg",
                  }
                }
              />,
              <ShopItem
                product={
                  Object {
                    "desc": "glossy giclee full colour",
                    "id": 6,
                    "price": 3.95,
                    "shopUrl": "desert-portrait",
                    "soldOut": true,
                    "title": "Desert Portrait",
                    "url": "shop06.jpg",
                  }
                }
              />,
            ],
            "className": "shop-container",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": "1",
              "nodeType": "function",
              "props": Object {
                "product": Object {
                  "desc": "monochrome screenprint",
                  "id": 1,
                  "price": 9.95,
                  "shopUrl": "mushroomjack",
                  "soldOut": false,
                  "title": "Mushroom Lumberjack",
                  "url": "shop01.jpg",
                },
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": "2",
              "nodeType": "function",
              "props": Object {
                "product": Object {
                  "desc": "five colour screenprint",
                  "id": 2,
                  "price": 22.5,
                  "shopUrl": "patterned-beetle",
                  "soldOut": false,
                  "title": "Patterned Beetle",
                  "url": "shop02.jpg",
                },
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": "3",
              "nodeType": "function",
              "props": Object {
                "product": Object {
                  "desc": "glossy giclee full colour",
                  "id": 3,
                  "price": 3.95,
                  "shopUrl": "desert-smoker",
                  "soldOut": false,
                  "title": "Desert Smoker",
                  "url": "shop03.jpg",
                },
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": "4",
              "nodeType": "function",
              "props": Object {
                "product": Object {
                  "desc": "art stock giclee",
                  "id": 4,
                  "price": 12.95,
                  "shopUrl": "robot-workshop",
                  "soldOut": false,
                  "title": "Robot Workshop",
                  "url": "shop04.jpg",
                },
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": "5",
              "nodeType": "function",
              "props": Object {
                "product": Object {
                  "desc": "five colour screenprint",
                  "id": 5,
                  "price": 22.5,
                  "shopUrl": "patterned-butterfly",
                  "soldOut": false,
                  "title": "Patterned Butterfly",
                  "url": "shop05.jpg",
                },
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": "6",
              "nodeType": "function",
              "props": Object {
                "product": Object {
                  "desc": "glossy giclee full colour",
                  "id": 6,
                  "price": 3.95,
                  "shopUrl": "desert-portrait",
                  "soldOut": true,
                  "title": "Desert Portrait",
                  "url": "shop06.jpg",
                },
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <h2>
                Be the first to know about new items
              </h2>,
              <div
                className="mailer-form"
              >
                <input
                  onChange={[Function]}
                  placeholder="your email"
                  type="email"
                  value=""
                />
                <button
                  onClick={[Function]}
                >
                  join
                </button>
              </div>,
            ],
            "className": "mailer-cta",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Be the first to know about new items",
              },
              "ref": null,
              "rendered": "Be the first to know about new items",
              "type": "h2",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  null,
                  <input
                    onChange={[Function]}
                    placeholder="your email"
                    type="email"
                    value=""
                  />,
                  <button
                    onClick={[Function]}
                  >
                    join
                  </button>,
                ],
                "className": "mailer-form",
              },
              "ref": null,
              "rendered": Array [
                null,
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "onChange": [Function],
                    "placeholder": "your email",
                    "type": "email",
                    "value": "",
                  },
                  "ref": null,
                  "rendered": null,
                  "type": "input",
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": "join",
                    "onClick": [Function],
                  },
                  "ref": null,
                  "rendered": "join",
                  "type": "button",
                },
              ],
              "type": "div",
            },
          ],
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
